pkg/disttask/framework/taskexecutor/task_executor.go

```diff
@@ -147,21 +150,84 @@ func (*BaseTaskExecutor) Init(_ context.Context) error {
 	return nil
 }
 
+// Run implements the TaskExecutor interface.
+func (e *BaseTaskExecutor) Run(resource *proto.StepResource) {
+	var err error
+	// task executor occupies resources, if there's no subtask to run for 10s,
+	// we release the resources so that other tasks can use them.
+	// 300ms + 600ms + 1.2s + 2s * 4 = 10.1s
+	backoffer := backoff.NewExponential(defaultCheckInterval, 2, maxCheckInterval)
+	checkInterval, noSubtaskCheckCnt := defaultCheckInterval, 0
+	for {
+		select {
+		case <-e.ctx.Done():
+			return
+		case <-time.After(checkInterval):
+		}
+		failpoint.Inject("mockStopManager", func() {
+			TestContexts.Store(e.id, &TestContext{make(chan struct{}), atomic.Bool{}})
+			go func() {
+				v, ok := TestContexts.Load(e.id)
+				if ok {
+					<-v.(*TestContext).TestSyncSubtaskRun
+					infosync.MockGlobalServerInfoManagerEntry.DeleteByExecID(e.id)
+				}
+			}()
+		})
+		if err = e.refreshTask(); err != nil {
+			if errors.Cause(err) == storage.ErrTaskNotFound {
+				return
+			}
+			e.logger.Error("refresh task failed", zap.Error(err))
+			continue
+		}
+		task := e.task.Load()
+		if task.State != proto.TaskStateRunning && task.State != proto.TaskStateReverting {
+			return
+		}
+		if exist, err := e.taskTable.HasSubtasksInStates(e.ctx, e.id, task.ID, task.Step,
+			unfinishedSubtaskStates...); err != nil {
+			e.logger.Error("check whether there are subtasks to run failed", zap.Error(err))
+			continue
+		} else if !exist {
+			if noSubtaskCheckCnt >= maxChecksWhenNoSubtask {
+				break
```

Comment:
Maybe add a log to indicate that this executor exits because there is no subtask to run for 10s.